dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/deploy/DefaultApplicationDeployer.java

```diff
@@ -358,15 +365,33 @@ private void initMetricsService() {
         metricsServiceExporter.init();
     }
 
+    @SuppressWarnings({"rawtypes"})
     private void initMetricsReporter() {
-        DefaultMetricsCollector collector =
-            applicationModel.getFrameworkModel().getBeanFactory().getOrRegisterBean(DefaultMetricsCollector.class);
+
+        ScopeBeanFactory beanFactory = applicationModel.getFrameworkModel().getBeanFactory();
         MetricsConfig metricsConfig = configManager.getMetrics().orElse(null);
+        this.eventMulticaster = beanFactory.getOrRegisterBean(SimpleMetricsEventMulticaster.class);
+
         // TODO compatible with old usage of metrics, remove protocol check after new metrics is ready for use.
         if (metricsConfig != null && PROTOCOL_PROMETHEUS.equals(metricsConfig.getProtocol())) {
+
+            DefaultMetricsCollector collector =
+                beanFactory.getOrRegisterBean(DefaultMetricsCollector.class);
             collector.setCollectEnabled(true);
-            collector.addApplicationInfo(applicationModel.getApplicationName(), Version.getVersion());
+            eventMulticaster.setAvailable();
+
+            List<MetricsListener> metricsListeners = applicationModel.getExtensionLoader(MetricsListener.class)
+                .getActivateExtensions();
+            metricsListeners.forEach(this.eventMulticaster::addListener);
+
+            List<MetricsCollector> customizeCollectors = applicationModel.getExtensionLoader(MetricsCollector.class)
+                .getActivateExtensions();
+            for (MetricsCollector customizeCollector : customizeCollectors) {
+                beanFactory.registerBean(customizeCollector);
+            }
```

Comment:
Refactor `MetricsListener` and `MetricsCollector` to scope bean. If there is more that one implmentation, introduce a factory to hold `MetricsListener` and `MetricsCollector`. Depends on `tExtensionLoader`'s siglton may cause some unexpected result.